drivers/stepper/stepper_motion_controllers/zephyr_stepper_motion_controller.c

```
@@ -0,0 +1,405 @@
+	}
+}
+
+static int z_stepper_motion_control_set_event_callback(const struct device *dev,
+						       stepper_event_callback_t cb, void *user_data)
+{
+	struct stepper_motion_control_data *data = dev->data;
+
+	K_SPINLOCK(&data->lock) {
+		data->callback = cb;
+	}
+	data->event_cb_user_data = user_data;
+	return 0;
+}
+
+void stepper_handle_timing_signal(const struct device *dev)
+{
+	struct stepper_motion_control_data *data = dev->data;
+
+	K_SPINLOCK(&data->lock) {
+		switch (data->run_mode) {
+		case STEPPER_RUN_MODE_POSITION:
+			position_mode_task(dev);
+			break;
+		case STEPPER_RUN_MODE_VELOCITY:
+			velocity_mode_task(dev);
+			break;
+		default:
+			LOG_WRN("Unsupported run mode: %d", data->run_mode);
+			break;
+		}
+	}
+}
+
+static int z_stepper_motion_control_enable(const struct device *dev)
+{
+	const struct stepper_motion_control_config *config = dev->config;
+
+	return stepper_enable(config->stepper);
+
+	return 0;
+}
+
+static int z_stepper_motion_control_disable(const struct device *dev)
+{
+	const struct stepper_motion_control_config *config = dev->config;
+
+	return stepper_disable(config->stepper);
+}
+
+static int z_stepper_motion_control_set_micro_step_res(const struct device *dev,
+						       const enum stepper_micro_step_resolution res)
+{
+	const struct stepper_motion_control_config *config = dev->config;
+
+	return stepper_set_micro_step_res(config->stepper, res);
+}
+
+static int z_stepper_motion_control_get_micro_step_res(const struct device *dev,
+						       enum stepper_micro_step_resolution *res)
+{
+	const struct stepper_motion_control_config *config = dev->config;
+
+	int ret = stepper_get_micro_step_res(config->stepper, res);
+
+	return ret;
+}
+
+static int stepper_motion_control_init(const struct device *dev)
+{
+	const struct stepper_motion_control_config *config = dev->config;
+	struct stepper_motion_control_data *data = dev->data;
+	int ret;
+
+	data->dev = dev;
+
+	if (config->timing_source->init) {
```

we should return an error if `config->timing_source->init` evaluates to `false` because in this case the controller wont be usable (and we have the worker timing source as fallback for that reason).

---

resolved